dataflow/flex-templates/pipeline_with_dependencies/README.md

```diff
@@ -0,0 +1,169 @@
+## Create an Artifact Registry repository
+
+```sh
+export REPOSITORY="your-repository"
+gcloud artifacts repositories create $REPOSITORY \
+    --repository-format=docker \
+    --location=$REGION
+
+gcloud artifacts repositories create $REPOSITORY \
+    --repository-format=docker \
+    --location=$REGION \
+    --project $PROJECT
+```
+
+## Build a Docker image for the pipeline runtime environment
+
+Using a [custom SDK container image](https://cloud.google.com/dataflow/docs/guides/using-custom-containers)
+allows flexible customizations of the runtime environment.
+
+We use the custom container image both to preinstall all of the pipeline dependencies before job submission and to create a reproducible runtime environment.
+
+To illustrate customizations, we use a [custom base base image](https://cloud.google.com/dataflow/docs/guides/build-container-image#use_a_custom_base_image) to build the SDK container image.
+
+```sh
+# Use a unique tag to version artifacts we build.
+export TAG=`date +%Y%m%d-%H%M%S`
+export SDK_CONTAINER_IMAGE="$REGION-docker.pkg.dev/$PROJECT/$REPOSITORY/my_base_image:$TAG"
+
+gcloud builds submit .  --tag $SDK_CONTAINER_IMAGE --project $PROJECT
+```
+
+You must rebuild the base image when your pipeline dependencies change.
+
+## Build the Flex Template
+
+To build the Flex Templates, use the [runtime image as the base image](https://cloud.google.com/dataflow/docs/guides/templates/configuring-flex-templates#use_custom_container_images).
+Using the runtime image as the base image allows us to reduce the number of Docker images that need to be maintained.
+It also ensures that the pipeline uses the same dependencies at submission and at runtime.
+
+```sh
+export TEMPLATE_FILE=gs://$BUCKET/longest-word-$TAG.json
+export TEMPLATE_IMAGE=$REGION-docker.pkg.dev/$PROJECT/$REPOSITORY/my_template_image:$TAG
+```
+
+```sh
+gcloud dataflow flex-template build $TEMPLATE_FILE  \
+    --image-gcr-path $TEMPLATE_IMAGE \
+    --sdk-language "PYTHON" \
+    --flex-template-base-image $SDK_CONTAINER_IMAGE \
+    --py-path "."  \
+    --env "FLEX_TEMPLATE_PYTHON_PY_FILE=main.py" \
+    --env "FLEX_TEMPLATE_PYTHON_SETUP_FILE=setup.py" \
+    --project $PROJECT
+```
+
+Note: Older versions of `gcloud` might not support Artifact Registry in the `--flex-template-base-image` option. You might need to upgrade `gcloud` or build the Flex Template from a `Dockerfile` using the instructions below.
+
+If you prefer to have a separate file to configure and build the Flex Template image (included in the example), such as a `Dockefile.flex_template` file, you can build the Flex Template as follows:
+
+```sh
+gcloud builds submit --config flex_template_cloudbuild.yaml \
+    --substitutions="_TEMPLATE_IMAGE=${TEMPLATE_IMAGE},_BASE_IMAGE=${SDK_CONTAINER_IMAGE}" \
+    --project $PROJECT
+
+gcloud dataflow flex-template build $TEMPLATE_FILE  \
+    --image $TEMPLATE_IMAGE \
+    --sdk-language "PYTHON" \
+    --project $PROJECT
+```
+
+## Run the template
```

Comment:
I suggest we could add a new optional section to check the container to verify the installed packages.